cmd/manager/main.go

```diff
@@ -105,96 +137,105 @@ func main() {
 	})
 	go http.ListenAndServe(":8080", nil)
 
-	// Wait until this instance becomes the leader.
-	logrus.Info("Waiting to become leader.")
-	err = leader.Become(context.TODO(), leaderElectionConfigMapName)
-	if err != nil {
-		logrus.Fatal(err, "Failed to retry for leader lock")
-	}
-	logrus.Info("Elected leader.")
+	ctx := signals.Context()
+	stopCh := ctx.Done()
 
-	logrus.Info("Starting the Cmd.")
+	leaderCtx, leaderCancel := context.WithCancel(context.Background())
+	defer leaderCancel()
 
-	// Handle the defaults
-	err = ensureDefaults(cfg, mgr.GetScheme())
-	if err != nil {
-		logrus.Fatal(err)
-	}
+	run := func(ctx context.Context) {
+		logger.Info("registering components")
+		var statusReporter status.Reporter = &status.NoOpReporter{}
+		if clusterOperatorName != "" {
+			statusReporter, err = status.NewReporter(cfg, mgr, namespace, clusterOperatorName, os.Getenv("RELEASE_VERSION"), stopCh)
+			if err != nil {
+				logger.Fatal(err)
+			}
+		}
+
+		// Populate the global default OperatorSources definition and config
+		if err := defaults.PopulateGlobals(); err != nil {
+			logger.Fatal(err)
+		}
 
-	// statusReportingDoneCh will be closed after the operator has successfully stopped reporting ClusterOperator status.
-	statusReportingDoneCh := statusReporter.StartReporting()
+		logger.Info("setting up controllers")
+		if err := controller.AddToManager(mgr, options.ControllerOptions{}); err != nil {
+			logger.Fatal(err)
+		}
 
-	// Start the Cmd
-	err = mgr.Start(stopCh)
+		if err := ensureDefaults(cfg, mgr.GetScheme()); err != nil {
+			logger.Fatalf("failed to setup the default catalogsource manifests: %v", err)
+		}
 
-	// Wait for ClusterOperator status reporting routine to close the statusReportingDoneCh channel.
-	<-statusReportingDoneCh
+		logger.Info("starting manager")
+		if err := mgr.Start(stopCh); err != nil {
+			logger.WithError(err).Error("unable to run manager")
+		}
 
-	exit(err)
-}
+		// statusReportingDoneCh will be closed after the operator has successfully stopped reporting ClusterOperator status.
+		statusReportingDoneCh := statusReporter.StartReporting()
+		// Wait for ClusterOperator status reporting routine to close the statusReportingDoneCh channel.
+		<-statusReportingDoneCh
+	}
 
-// exit stops the reporting of ClusterOperator status and exits with the proper exit code.
-func exit(err error) {
-	// If an error exists then exit with status set to 1
+	client, err := kubernetes.NewForConfig(mgr.GetConfig())
 	if err != nil {
-		logrus.Fatalf("The operator encountered an error, exit code 1: %v", err)
+		logger.Fatal(fmt.Errorf("failed to initialize the kubernetes clientset: %v", err))
 	}
 
-	// No error, graceful termination
-	logrus.Info("The operator exited gracefully, exit code 0")
-	os.Exit(0)
+	id := os.Getenv("POD_NAME")
+	if id == "" {
+		logger.Info("failed to determine $POD_NAME falling back to hostname")
+		id, err = os.Hostname()
+		if err != nil {
+			logger.Fatal(err)
+		}
+	}
```

Comment:
Just a quick note about this behavior - I was reading the operator-lib's leader election package [godoc](https://github.com/operator-framework/operator-lib/blob/main/leader/doc.go#L44-L52) and saw the following was documented:

```
Leader for Life requires that all candidate Pods be in the same Namespace. It
uses the downwards API to determine the pod name, as hostname is not reliable.
You should run it configured with:
env:
  - name: POD_NAME
    valueFrom:
      fieldRef:
        fieldPath: metadata.name
```

We already specify the POD_NAME in the marketplace-operator deployment, so we _should_ have access to that value, and if not, fall back to getting this value by querying for pod hostname. I was mainly testing this on a crc cluster (e.g. single node) and ran into some of those inconsistencies between os.Hostname() and pod metadata.Name values.